design/kata-2-0-metrics.md

```diff
@@ -0,0 +1,1343 @@
+# Kata 2.0 Metrics Design
+
+* [Limitations of Kata 1.x and the target of Kata 2.0](#limitations-of-kata-1x-and-the-target-of-kata-20)
+* [Metrics architecture](#metrics-architecture)
+  * [Kata magent](#kata-magent)
+  * [Kata runtime](#kata-runtime)
+  * [Kata agent](#kata-agent)
+  * [Performance and overhead](#performance-and-overhead)
+* [Metrics list](#metrics-list)
+  * [Kata agent metrics](#kata-agent-metrics)
+  * [Kata guest OS](#kata-guest-os)
+  * [Hypervisor metrics](#hypervisor-metrics)
+  * [Kata magent metrics](#kata-magent-metrics)
+  * [Kata shim metrics](#kata-shim-metrics)
+
+
+Kata implement CRI's API and support [`ContainerStats`](https://github.com/kubernetes/kubernetes/blob/release-1.18/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto#L101) and [`ListContainerStats`](https://github.com/kubernetes/kubernetes/blob/release-1.18/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto#L103) interfaces to expose containers metrics. User can use these interface to get basic metrics about container.
+
+But unlike `runc`, Kata is a VM-based runtime and has a distinguish architecture.
+
+## Limitations of Kata 1.x and the target of Kata 2.0
+
+There are some limitations, or called missing parts, the lack of observability may become an obstacles to run Kata containers in large scale.
+
+In Kata 2.0, these parts will be took to provide more details about the system.
+
+- Shim process(Kata runtime)
+- Hypervisor stats
+- Agent process
+- Guest OS stats
+
+## Metrics architecture
+
+Kata 2.0 metrics strongly depend on [Prometheus](https://prometheus.io/), a graduated project from CNCF.
+
+The new design will introduce a new command(component) named `kata-magent`, the `m` of `magent` means metrics or management, this command is used to manage/monitor Kata runtime it self other than containers. It's the management interface with Kata runtime, and we can do something like these:
+
+- Get metrics
+- Get events
+- Extend hypervisor commands
+
+In this document we will cover metrics only. And until now it only supports metrics function.
+
+This is the architecture overview metrics in Kata Containers 2.0.
+
+![Kata Containers 2.0 metrics](arch-images/kata-2-metrics.png)
+
+
+And the sequence diagram is shown below:
+
+![Kata Containers 2.0 metrics ](arch-images/kata-metrics-sequence-diagram.png)
+
+
+### Kata magent
+
+`magent` is a management agent on one nodes, where many Kata containers are running. `magent`'s work include:
```

Comment:
I prefer `kata-exporter`/`kata-collector`/`kata-inspector`, but I also think we need more advices. @bergwolf

Reply:
It's worth pointing out that I've raised https://github.com/kata-containers/kata-containers/pull/225 to add a new "trace forwarder" component for tracing.

*Potentially* the "`magent`" (https://github.com/kata-containers/kata-containers/pull/297) component could also handle this task.

I'm just highlighting this as there is some amount of similarity. However, I'm not really convinved we should combine them since:

- I've already written the trace forwarder PR, and the new code is in rust, rather than golang.
- It would break the design slightly of "`magent`" since my component listens to the agent via VSOCK, not to the shim v2 process.